addons/portal/models/mail_thread.py

```diff
@@ -87,3 +91,43 @@ def _portal_get_parent_hash_token(self, pid):
         :return: False or logical parent's _sign_token() result
         """
         return False
+
+    def _get_portal_partner(self, pid=False):
+        req = request or wsrequest
+        partner = self.env.user.partner_id
+        if pid:
+            partner = self.env["res.partner"].browse(pid)
+        else:
+            if request.env.user._is_public() and hasattr(self, "partner_id") and self.partner_id:
+                partner = self.partner_id
+            elif not partner:
+                raise NotFound()
+        req.update_context(portal_partner=partner)
+        if hasattr(self, "env"):
+            self.env.context = frozendict({**self.env.context, "portal_partner": partner})
+        return partner
+
+    @api.model
+    def _get_thread_with_access(self, thread_id, **kwargs):
+        token, _hash, pid = (kwargs.get(key) for key in ["token", "hash", "pid"])
+        if token or (_hash and pid):
+            has_access = False
+            record = self.browse(thread_id).sudo()
+            if _hash and pid:  # Signed Token Case: hash implies token is signed by partner pid
+                pid = int(pid)
+                has_access = consteq(_hash, record._sign_token(pid))
+                if not has_access:
+                    parent_sign_token = record._portal_get_parent_hash_token(pid)
+                    has_access = parent_sign_token and consteq(_hash, parent_sign_token)
+            elif token:  # Token Case: token is the global one of the document
+                token_field = self._mail_post_token_field
+                has_access = token and record and consteq(record[token_field], token)
+            if not has_access:
+                raise Forbidden()
```

Comment:
I don't like that a model method returns a http error, a get thread method should simply return no thread if it can't be accessed or can't be found.

Called should be responsibly of handing the missing thread accordingly.

This should probably fallback to returning the result of super instead.